### PR TITLE
LPS-174906 Add support for file upload through the API Explorer page

### PR DESCRIPTION
**Pull request by jalber786**
**What's new?:**
- Added the possibility of uploading files in endpoints with Multipart/form-data media types in API Explorer.
- Modified RESTBuilder code to allow for the autogeneration of the necessary annotations.
- Modified the base-resource.ftl to use the RESTBuilder tools generated for the APIBuilder.

See the following **Jira ticket**: [LPS-174906](https://liferay.atlassian.net/browse/LPS-174906)

### CODE REVIEW

**Comment by jalber786** [comment]
ci:test:sf

**Comment by jalber786** [comment]
ci:test:relevant

File: modules/util/portal-tools-rest-builder/src/main/java/com/liferay/portal/tools/rest/builder/internal/freemarker/tool/java/parser/ResourceOpenAPIParser.java [hunk 1]
```diff
@@ -146,13 +146,54 @@ public static String getMethodAnnotations(
...
+					sb.append("RequestBody.class)))");
+				}
 			}
 
 			sb.append(")");
 
 			methodAnnotations.add(sb.toString());
 		}
 
+		if ((operation.getDescription() == null) &&
```
Comment (sergiojimcos) [inline modules/util/portal-tools-rest-builder/src/main/java/com/liferay/portal/tools/rest/builder/internal/freemarker/tool/java/parser/ResourceOpenAPIParser.java:174]: Why are you checking if `operation.getDescription()` is null? It is managed in the above part, exactly in line 146. Maybe I am missunderstanding something :thinking:

File: modules/util/portal-tools-rest-builder/src/main/java/com/liferay/portal/tools/rest/builder/internal/freemarker/tool/java/parser/ResourceOpenAPIParser.java [hunk 2]
```diff
@@ -146,13 +146,54 @@ public static String getMethodAnnotations(
...
+			javaMethodSignature.getRequestBodyMediaTypes(
+			).contains(
+				"multipart/form-data"
+			) && !operation.isDeprecated() &&
+			(getMultipartBodySchemas(javaMethodSignature) != null)) {
```
Comment (sergiojimcos) [inline modules/util/portal-tools-rest-builder/src/main/java/com/liferay/portal/tools/rest/builder/internal/freemarker/tool/java/parser/ResourceOpenAPIParser.java:179]: I think that can simplify this condition
```java
javaMethodSignature.getRequestBodyMediaTypes(
			).contains(
				"multipart/form-data"
			) 
```
inside of `getMultipartBodySchemas`. something like:

```java
public static Map<String, Schema> getMultipartBodySchemas(
		JavaMethodSignature javaMethodSignature) {

		Operation operation = javaMethodSignature.getOperation();

		RequestBody requestBody = operation.getRequestBody();

		Map<String, Content> contentMap = requestBody.getContent();

		Content multipartContent = contentMap.get("multipart/form-data");
                
                if (content!=null) {
                           Schema schema = multipartContent.getSchema();

		          return schema.getPropertySchemas();

                }
                

		return null;
	}
```

**Review by sergiojimcos — CHANGES_REQUESTED** [review]
Review some comments

**Comment by jalber786** [comment]
ci:test:sf

**Comment by jalber786** [comment]
ci:test:relevant

**Comment by jalber786** [comment]
ci:stop

**Comment by jalber786** [comment]
ci:test:sf

**Comment by jalber786** [comment]
ci:test:relevant

**Comment by jalber786** [comment]
ci:test:relevant

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl [hunk 3]
```diff
@@ -1371,6 +1376,26 @@ public abstract class Base${schemaName}ResourceImpl
 
 	private static final com.liferay.portal.kernel.log.Log _log = LogFactoryUtil.getLog(Base${schemaName}ResourceImpl.class);
 
+	<#if generateMultipartBodyClasses?has_content>
+		<#list generateMultipartBodyClasses as javaMethodSignatureWithMultipartBody>
+			private class _${stringUtil.upperCaseFirstLetter(javaMethodSignatureWithMultipartBody.methodName)}RequestBody{
```
Comment (4lejandrito) [inline modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl:1381]: Private classes don't need to be prefixed by `_`.

**Review by 4lejandrito — CHANGES_REQUESTED** [review]
Please also make sure you manually test the APIs that have been changed:

image

**Comment by jalber786** [comment]
ci:test:sf

**Comment by jalber786** [comment]
ci:test:relevant

**Review by sergiojimcos — APPROVED** [review]

**Comment by jalber786** [comment]
ci:test:relevant

**Comment by yanan** [comment]
Failures are not related, here is the [compare run results](https://testray.liferay.com/home/-/testray/runs/compare?compare=true&view=details&statusA=3&testrayRunIdB=2422541312&testrayRunIdA=2423040066&statusB=2)
- the only failure `modules-semantic-versioning-jdk8/0/0` passed in the second run

**Comment by jalber786** [comment]
ci:forward:force

**Comment by jalber786** [comment]
Manually forwarded through https://github.com/brianchandotcom/liferay-portal/pull/140655

**Pull request closed without merging**